Out of bounds nanosecond timestamp: 1-01-01 00:00:00

Good afternoon.

I have a table where there is a column named maxDate. 

There are some rows (like 1%) that can have the following date: 0001-01-01T00:00:00.

The problem is that it lives in a GBQ from a external vendor to the company, so icannot modify the data and when i call th query like this:
    df = pandas_gbq.read_gbq(
        sql, credentials=credentials)

I have the error:  File "pandas\_libs\tslibs\np_datetime.pyx", line 117, in pandas._libs.tslibs.np_datetime.check_dts_bounds
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 1-01-01 00:00:00.

Is there a workaround or config so i can execute the query correctly?

Thank you in advance.

I can reproduce this in a test in #442 

```
________________________ test_dataframe_round_trip_with_table_schema[load_parquet-issue365-extreme-datetimes] _________________________

method_under_test = functools.partial(<function to_gbq at 0x7f9090ebb710>, project_id='swast-scratch', credentials=<google.oauth2.credentials.Credentials object at 0x7f905058b250>)
random_dataset_id = 'python_bigquery_pandas_tests_system_20211206201607_aa0fdb'
bigquery_client = <google.cloud.bigquery.client.Client object at 0x7f9050574950>
input_df =    row_num    date_col                datetime_col                     timestamp_col
0        1  0001-01-01         00...         1970-01-01 00:00:00+00:00
2        3  9999-12-31  9999-12-31 23:59:59.999999  9999-12-31 23:59:59.999999+00:00
expected_df =    row_num    date_col                datetime_col                     timestamp_col
0        1  0001-01-01         00...         1970-01-01 00:00:00+00:00
2        3  9999-12-31  9999-12-31 23:59:59.999999  9999-12-31 23:59:59.999999+00:00
table_schema = [{'name': 'row_num', 'type': 'INTEGER'}, {'name': 'date_col', 'type': 'DATE'}, {'name': 'datetime_col', 'type': 'DATETIME'}, {'name': 'timestamp_col', 'type': 'TIMESTAMP'}]
api_method = 'load_parquet', api_methods = {'load_csv', 'load_parquet'}

                                59,
                                59,
                                999999,
                                tzinfo=datetime.timezone.utc,
                            ),
                        ],
                    },
                    columns=["row_num", "date_col", "datetime_col", "timestamp_col"],
                ),
                table_schema=[
                    {"name": "row_num", "type": "INTEGER"},
                    {"name": "date_col", "type": "DATE"},
                    {"name": "datetime_col", "type": "DATETIME"},
                    {"name": "timestamp_col", "type": "TIMESTAMP"},
                ],
            ),
            id="issue365-extreme-datetimes",
        ),
    ]
    
    
    @pytest.mark.parametrize(
>       ["input_df", "expected_df", "table_schema", "api_methods"], DATAFRAME_ROUND_TRIPS
    )
    def test_dataframe_round_trip_with_table_schema(
        method_under_test,
        random_dataset_id,
        bigquery_client,
        input_df,
        expected_df,
        table_schema,
        api_method,
        api_methods,
    ):

tests/system/test_to_gbq.py:243: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.nox/system-3-7/lib/python3.7/site-packages/google/cloud/bigquery/table.py:1729: in to_dataframe
    df = record_batch.to_pandas(date_as_object=date_as_object)
pyarrow/array.pxi:751: in pyarrow.lib._PandasConvertible.to_pandas
    ???
pyarrow/table.pxi:1668: in pyarrow.lib.Table._to_pandas
    ???
.nox/system-3-7/lib/python3.7/site-packages/pyarrow/pandas_compat.py:792: in table_to_blockmanager
    blocks = _table_to_blocks(options, table, categories, ext_columns_dtypes)
.nox/system-3-7/lib/python3.7/site-packages/pyarrow/pandas_compat.py:1132: in _table_to_blocks
    list(extension_columns.keys()))
pyarrow/table.pxi:1117: in pyarrow.lib.table_to_blocks
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pyarrow.lib.ArrowInvalid: Casting from timestamp[us] to timestamp[ns] would result in out of bounds timestamp: -62135596800000000

pyarrow/error.pxi:84: ArrowInvalid
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------
100%|██████████| 1/1 [00:00<00:00, 5915.80it/s]
__________________________ test_dataframe_round_trip_with_table_schema[load_csv-issue365-extreme-datetimes] ___________________________

method_under_test = functools.partial(<function to_gbq at 0x7f9090ebb710>, project_id='swast-scratch', credentials=<google.oauth2.credentials.Credentials object at 0x7f905058b250>)
random_dataset_id = 'python_bigquery_pandas_tests_system_20211206201653_cf3865'
bigquery_client = <google.cloud.bigquery.client.Client object at 0x7f9050574950>
input_df =    row_num    date_col                datetime_col                     timestamp_col
0        1  0001-01-01         00...         1970-01-01 00:00:00+00:00
2        3  9999-12-31  9999-12-31 23:59:59.999999  9999-12-31 23:59:59.999999+00:00
expected_df =    row_num    date_col                datetime_col                     timestamp_col
0        1  0001-01-01         00...         1970-01-01 00:00:00+00:00
2        3  9999-12-31  9999-12-31 23:59:59.999999  9999-12-31 23:59:59.999999+00:00
table_schema = [{'name': 'row_num', 'type': 'INTEGER'}, {'name': 'date_col', 'type': 'DATE'}, {'name': 'datetime_col', 'type': 'DATETIME'}, {'name': 'timestamp_col', 'type': 'TIMESTAMP'}]
api_method = 'load_csv', api_methods = {'load_csv', 'load_parquet'}

                                59,
                                59,
                                999999,
                                tzinfo=datetime.timezone.utc,
                            ),
                        ],
                    },
                    columns=["row_num", "date_col", "datetime_col", "timestamp_col"],
                ),
                table_schema=[
                    {"name": "row_num", "type": "INTEGER"},
                    {"name": "date_col", "type": "DATE"},
                    {"name": "datetime_col", "type": "DATETIME"},
                    {"name": "timestamp_col", "type": "TIMESTAMP"},
                ],
            ),
            id="issue365-extreme-datetimes",
        ),
    ]
    
    
    @pytest.mark.parametrize(
>       ["input_df", "expected_df", "table_schema", "api_methods"], DATAFRAME_ROUND_TRIPS
    )
    def test_dataframe_round_trip_with_table_schema(
        method_under_test,
        random_dataset_id,
        bigquery_client,
        input_df,
        expected_df,
        table_schema,
        api_method,
        api_methods,
    ):

tests/system/test_to_gbq.py:243: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.nox/system-3-7/lib/python3.7/site-packages/google/cloud/bigquery/table.py:1729: in to_dataframe
    df = record_batch.to_pandas(date_as_object=date_as_object)
pyarrow/array.pxi:751: in pyarrow.lib._PandasConvertible.to_pandas
    ???
pyarrow/table.pxi:1668: in pyarrow.lib.Table._to_pandas
    ???
.nox/system-3-7/lib/python3.7/site-packages/pyarrow/pandas_compat.py:792: in table_to_blockmanager
    blocks = _table_to_blocks(options, table, categories, ext_columns_dtypes)
.nox/system-3-7/lib/python3.7/site-packages/pyarrow/pandas_compat.py:1132: in _table_to_blocks
    list(extension_columns.keys()))
pyarrow/table.pxi:1117: in pyarrow.lib.table_to_blocks
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pyarrow.lib.ArrowInvalid: Casting from timestamp[us] to timestamp[ns] would result in out of bounds timestamp: -62135596800000000

pyarrow/error.pxi:84: ArrowInvalid
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------
1it [00:04,  4.31s/it]
========================================================== warnings summary ===========================================================
tests/system/test_gbq.py::TestReadGBQIntegration::test_array_length_zero
  /Users/swast/src/github.com/googleapis/python-bigquery-pandas/.nox/system-3-7/lib/python3.7/site-packages/pandas/core/dtypes/missing.py:415: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
    if left_value != right_value:

tests/system/test_gbq.py::TestToGBQIntegration::test_upload_data
  /Users/swast/src/github.com/googleapis/python-bigquery-pandas/tests/system/test_gbq.py:948: PendingDeprecationWarning: chunksize will be ignored when using api_method='load_csv' in a future version of pandas-gbq
    credentials=self.credentials,

tests/system/test_gbq.py::TestToGBQIntegration::test_upload_data_if_table_exists_append
  /Users/swast/src/github.com/googleapis/python-bigquery-pandas/tests/system/test_gbq.py:1032: PendingDeprecationWarning: chunksize will be ignored when using api_method='load_csv' in a future version of pandas-gbq
    credentials=self.credentials,

tests/system/test_gbq.py::TestToGBQIntegration::test_upload_subset_columns_if_table_exists_append
  /Users/swast/src/github.com/googleapis/python-bigquery-pandas/tests/system/test_gbq.py:1078: PendingDeprecationWarning: chunksize will be ignored when using api_method='load_csv' in a future version of pandas-gbq
    credentials=self.credentials,

tests/system/test_gbq.py::TestToGBQIntegration::test_upload_data_if_table_exists_replace
  /Users/swast/src/github.com/googleapis/python-bigquery-pandas/tests/system/test_gbq.py:1112: PendingDeprecationWarning: chunksize will be ignored when using api_method='load_csv' in a future version of pandas-gbq
    credentials=self.credentials,

tests/system/test_gbq.py::TestToGBQIntegration::test_upload_data_flexible_column_order
  /Users/swast/src/github.com/googleapis/python-bigquery-pandas/tests/system/test_gbq.py:1241: PendingDeprecationWarning: chunksize will be ignored when using api_method='load_csv' in a future version of pandas-gbq
    credentials=self.credentials,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
------------- generated xml file: /Users/swast/src/github.com/googleapis/python-bigquery-pandas/system_3.7_sponge_log.xml -------------
======================================================= short test summary info =======================================================
FAILED tests/system/test_to_gbq.py::test_dataframe_round_trip_with_table_schema[load_parquet-issue365-extreme-datetimes] - pyarrow.l...
FAILED tests/system/test_to_gbq.py::test_dataframe_round_trip_with_table_schema[load_csv-issue365-extreme-datetimes] - pyarrow.lib.A...
2 failed, 112 passed, 3 skipped, 6 warnings in 634.00s (0:10:34)
nox > Command py.test --quiet --junitxml=system_3.7_sponge_log.xml tests/system failed with exit code 1
nox > Session system-3.7 failed.
```

The good news is that it's fixed when I work with the `google-cloud-bigquery` v3 prerelease. https://github.com/googleapis/python-bigquery/pull/972 For other versions we'll need to catch and retry with `date_as_object=True`.

--- a/ci/requirements-3.7-0.24.2.conda
+++ b/ci/requirements-3.7-0.24.2.conda
@@ -1,10 +1,11 @@
 codecov
 coverage
-db-dtypes==0.3.0
+db-dtypes==0.3.1
 fastavro
 flake8
 numpy==1.16.6
-google-cloud-bigquery==1.11.1
+google-cloud-bigquery==1.26.1
+google-cloud-bigquery-storage==1.1.0
 pyarrow==3.0.0
 pydata-google-auth
 pytest

--- a/ci/requirements-3.7-0.24.2.conda
+++ b/ci/requirements-3.7-0.24.2.conda
@@ -4,8 +4,7 @@ db-dtypes==0.3.1
 fastavro
 flake8
 numpy==1.16.6
-google-cloud-bigquery==1.26.1
-google-cloud-bigquery-storage==1.1.0
+google-cloud-bigquery==1.11.1
 pyarrow==3.0.0
 pydata-google-auth
 pytest

--- a/pandas_gbq/load.py
+++ b/pandas_gbq/load.py
@@ -94,8 +94,13 @@ def cast_dataframe_for_parquet(
             # .astype() with DateDtype. With .astype(), I get the error:
             #
             # TypeError: Cannot interpret '<db_dtypes.DateDtype ...>' as a data type
-            cast_column = pandas.Series(
-                dataframe[column_name], dtype=db_dtypes.DateDtype()
+            cast_column = dataframe[column_name].astype(
+                dtype=db_dtypes.DateDtype(),
+                # Return the original column if there was an error converting
+                # to the dtype, such as is there is a date outside the
+                # supported range.
+                # https://github.com/googleapis/python-bigquery-pandas/issues/441
+                errors="ignore",
             )
         elif column_type in {"NUMERIC", "DECIMAL", "BIGNUMERIC", "BIGDECIMAL"}:
             cast_column = dataframe[column_name].map(decimal.Decimal)

--- a/pandas_gbq/load.py
+++ b/pandas_gbq/load.py
@@ -90,10 +90,6 @@ def cast_dataframe_for_parquet(
             # Use extension dtype first so that it uses the correct equality operator.
             and db_dtypes.DateDtype() != dataframe[column_name].dtype
         ):
-            # Construct converted column manually, because I can't use
-            # .astype() with DateDtype. With .astype(), I get the error:
-            #
-            # TypeError: Cannot interpret '<db_dtypes.DateDtype ...>' as a data type
             cast_column = dataframe[column_name].astype(
                 dtype=db_dtypes.DateDtype(),
                 # Return the original column if there was an error converting

--- a/setup.py
+++ b/setup.py
@@ -32,8 +32,7 @@ dependencies = [
     "google-auth-oauthlib",
     # 2.4.* has a bug where waiting for the query can hang indefinitely.
     # https://github.com/pydata/pandas-gbq/issues/343
-    "google-cloud-bigquery >=1.26.1,<3.0.0dev,!=2.4.*",
-    "google-cloud-bigquery-storage >=1.1.0,<3.0.0dev",
+    "google-cloud-bigquery[bqstorage,pandas] >=1.11.1,<3.0.0dev,!=2.4.*",
 ]
 extras = {
     "tqdm": "tqdm>=4.23.0",

--- a/setup.py
+++ b/setup.py
@@ -23,16 +23,17 @@ description = "Google BigQuery connector for pandas"
 release_status = "Development Status :: 4 - Beta"
 dependencies = [
     "setuptools",
-    "db-dtypes >=0.3.0,<2.0.0",
-    "numpy>=1.16.6",
-    "pandas>=0.24.2",
+    "db-dtypes >=0.3.1,<2.0.0",
+    "numpy >=1.16.6",
+    "pandas >=0.24.2",
     "pyarrow >=3.0.0, <7.0dev",
     "pydata-google-auth",
     "google-auth",
     "google-auth-oauthlib",
     # 2.4.* has a bug where waiting for the query can hang indefinitely.
     # https://github.com/pydata/pandas-gbq/issues/343
-    "google-cloud-bigquery[bqstorage,pandas]>=1.11.1,<3.0.0dev,!=2.4.*",
+    "google-cloud-bigquery >=1.26.1,<3.0.0dev,!=2.4.*",
+    "google-cloud-bigquery-storage >=1.1.0,<3.0.0dev",
 ]
 extras = {
     "tqdm": "tqdm>=4.23.0",

--- a/testing/constraints-3.7.txt
+++ b/testing/constraints-3.7.txt
@@ -5,10 +5,10 @@
 #
 # e.g., if setup.py has "foo >= 1.14.0, < 2.0.0dev",
 # Then this file should have foo==1.14.0
-db-dtypes==0.3.0
-google-auth==1.4.1
+db-dtypes==0.3.1
+google-auth==1.18.0
 google-auth-oauthlib==0.0.1
-google-cloud-bigquery==1.11.1
+google-cloud-bigquery==1.26.1
 google-cloud-bigquery-storage==1.1.0
 numpy==1.16.6
 pandas==0.24.2

--- a/testing/constraints-3.7.txt
+++ b/testing/constraints-3.7.txt
@@ -6,9 +6,9 @@
 # e.g., if setup.py has "foo >= 1.14.0, < 2.0.0dev",
 # Then this file should have foo==1.14.0
 db-dtypes==0.3.1
-google-auth==1.18.0
+google-auth==1.4.1
 google-auth-oauthlib==0.0.1
-google-cloud-bigquery==1.26.1
+google-cloud-bigquery==1.11.1
 google-cloud-bigquery-storage==1.1.0
 numpy==1.16.6
 pandas==0.24.2

--- a/tests/system/test_to_gbq.py
+++ b/tests/system/test_to_gbq.py
@@ -188,6 +188,54 @@ DATAFRAME_ROUND_TRIPS = [
             {"name": "num_col", "type": "NUMERIC"},
         ],
     ),
+    pytest.param(
+        *DataFrameRoundTripTestCase(
+            input_df=pandas.DataFrame(
+                {
+                    "row_num": [1, 2, 3],
+                    # DATE valuess outside the pandas range for timestamp
+                    # aren't supported by the db-dtypes package.
+                    # https://github.com/googleapis/python-bigquery-pandas/issues/441
+                    "date_col": [
+                        datetime.date(1, 1, 1),
+                        datetime.date(1970, 1, 1),
+                        datetime.date(9999, 12, 31),
+                    ],
+                    # DATETIME values outside of the range for pandas timestamp
+                    # require `date_as_object` parameter in
+                    # google-cloud-bigquery versions 1.x and 2.x.
+                    # https://github.com/googleapis/python-bigquery-pandas/issues/365
+                    "datetime_col": [
+                        datetime.datetime(1, 1, 1),
+                        datetime.datetime(1970, 1, 1),
+                        datetime.datetime(9999, 12, 31, 23, 59, 59, 999999),
+                    ],
+                    "timestamp_col": [
+                        datetime.datetime(1, 1, 1, tzinfo=datetime.timezone.utc),
+                        datetime.datetime(1970, 1, 1, tzinfo=datetime.timezone.utc),
+                        datetime.datetime(
+                            9999,
+                            12,
+                            31,
+                            23,
+                            59,
+                            59,
+                            999999,
+                            tzinfo=datetime.timezone.utc,
+                        ),
+                    ],
+                },
+                columns=["row_num", "date_col", "datetime_col", "timestamp_col"],
+            ),
+            table_schema=[
+                {"name": "row_num", "type": "INTEGER"},
+                {"name": "date_col", "type": "DATE"},
+                {"name": "datetime_col", "type": "DATETIME"},
+                {"name": "timestamp_col", "type": "TIMESTAMP"},
+            ],
+        ),
+        id="issue365-extreme-datetimes",
+    ),
 ]
 
 

--- a/tests/system/test_to_gbq.py
+++ b/tests/system/test_to_gbq.py
@@ -201,29 +201,29 @@ DATAFRAME_ROUND_TRIPS = [
                         datetime.date(1970, 1, 1),
                         datetime.date(9999, 12, 31),
                     ],
-                    # DATETIME values outside of the range for pandas timestamp
-                    # require `date_as_object` parameter in
+                    # TODO: DATETIME/TIMESTAMP values outside of the range for
+                    # pandas timestamp require `date_as_object` parameter in
                     # google-cloud-bigquery versions 1.x and 2.x.
                     # https://github.com/googleapis/python-bigquery-pandas/issues/365
-                    "datetime_col": [
-                        datetime.datetime(1, 1, 1),
-                        datetime.datetime(1970, 1, 1),
-                        datetime.datetime(9999, 12, 31, 23, 59, 59, 999999),
-                    ],
-                    "timestamp_col": [
-                        datetime.datetime(1, 1, 1, tzinfo=datetime.timezone.utc),
-                        datetime.datetime(1970, 1, 1, tzinfo=datetime.timezone.utc),
-                        datetime.datetime(
-                            9999,
-                            12,
-                            31,
-                            23,
-                            59,
-                            59,
-                            999999,
-                            tzinfo=datetime.timezone.utc,
-                        ),
-                    ],
+                    # "datetime_col": [
+                    #     datetime.datetime(1, 1, 1),
+                    #     datetime.datetime(1970, 1, 1),
+                    #     datetime.datetime(9999, 12, 31, 23, 59, 59, 999999),
+                    # ],
+                    # "timestamp_col": [
+                    #     datetime.datetime(1, 1, 1, tzinfo=datetime.timezone.utc),
+                    #     datetime.datetime(1970, 1, 1, tzinfo=datetime.timezone.utc),
+                    #     datetime.datetime(
+                    #         9999,
+                    #         12,
+                    #         31,
+                    #         23,
+                    #         59,
+                    #         59,
+                    #         999999,
+                    #         tzinfo=datetime.timezone.utc,
+                    #     ),
+                    # ],
                 },
                 columns=["row_num", "date_col", "datetime_col", "timestamp_col"],
             ),

